File: src/services/compliance/scoring.ts
import { prisma } from "@/lib/db";
import { Prisma } from "@prisma/client";

// ── Types ─────────────────────────────────────────────────────────────────

export interface ComplianceScoreResult {
  projectId: string;
  score: number; // 0-100
  compliancePercentage: number | null;
  onTimeCount: number;
  totalCount: number;
  missedCount: number;
  currentStreak: number;
  bestStreak: number;
  streakBrokenAt: Date | null;
  protectedClaimsValue: number;
  atRiskValue: number;
  atRiskCount: number;
  activeCount: number;
  upcomingCount: number;
  display: {
    scorePercentage: string;
    streakDisplay: string;
    protectedValue: string;
    verdict: string;
  };
}

// ── Score Calculator ──────────────────────────────────────────────────────

export async function calculateComplianceScore(
  projectId: string
): Promise<ComplianceScoreResult> {
  // Get all notices with delivery status
  const notices = await prisma.complianceNotice.findMany({
    where: {
      projectId,
      status: { in: ["SENT", "ACKNOWLEDGED", "EXPIRED"] },
    },
    orderBy: { sentAt: "asc" },
  });

  // Count on-time vs late vs expired
  const onTimeNotices = notices.filter((n) => n.onTimeStatus === true);
  const missedNotices = notices.filter(
    (n) => n.onTimeStatus === false || n.status === "EXPIRED"
  );

  const onTimeCount = onTimeNotices.length;
  const totalCount = notices.length;
  const missedCount = missedNotices.length;

  const compliancePercentage =
    totalCount > 0 ? (onTimeCount / totalCount) * 100 : null;
  const score =
    compliancePercentage !== null ? Math.round(compliancePercentage) : 100;

  // Calculate streak
  const { currentStreak, bestStreak, streakBrokenAt } =
    calculateStreak(notices);

  // Calculate protected claims value
  const protectedClaimsValue =
    await calculateProtectedClaimsValue(projectId);

  // Get at-risk deadlines (CRITICAL or WARNING severity)
  const atRiskCount = await prisma.complianceDeadline.count({
    where: {
      projectId,
      status: "ACTIVE",
      severity: { in: ["CRITICAL", "WARNING"] },
    },
  });

  // Get active deadlines
  const activeCount = await prisma.complianceDeadline.count({
    where: { projectId, status: "ACTIVE" },
  });

  // Get upcoming deadlines (next 14 days)
  const upcomingCount = await prisma.complianceDeadline.count({
    where: {
      projectId,
      status: "ACTIVE",
      calculatedDeadline: {
        lte: new Date(Date.now() + 14 * 24 * 60 * 60 * 1000),
      },
    },
  });

  // Estimate at-risk value from change events linked to at-risk deadlines
  const atRiskDeadlines = await prisma.complianceDeadline.findMany({
    where: {
      projectId,
      status: "ACTIVE",
      severity: { in: ["CRITICAL", "WARNING", "EXPIRED"] },
      triggerEventId: { not: null },
    },
    select: { triggerEventId: true },
  });

  let atRiskValue = 0;
  if (atRiskDeadlines.length > 0) {
    const eventIds = atRiskDeadlines
      .map((d) => d.triggerEventId)
      .filter(Boolean) as string[];
    if (eventIds.length > 0) {
      const events = await prisma.changeEvent.findMany({
        where: { id: { in: eventIds } },
        select: { estimatedValue: true },
      });
      atRiskValue = events.reduce(
        (sum, e) => sum + (e.estimatedValue ? Number(e.estimatedValue) : 0),
        0
      );
    }
  }

  // Persist score
  await prisma.complianceScore.create({
    data: {
      projectId,
      score,
      details: JSON.parse(
        JSON.stringify({
          compliancePercentage,
          onTimeCount,
          totalCount,
          missedCount,
        })
      ),
      currentStreak,
      bestStreak,
      streakBrokenAt,
      protectedClaimsValue: new Prisma.Decimal(protectedClaimsValue),
      atRiskValue: new Prisma.Decimal(atRiskValue),
      onTimeCount,
      totalCount,
      missedCount,
      atRiskCount,
      activeCount,
      upcomingCount,
      lastCalculatedAt: new Date(),
    },
  });

  return {
    projectId,
    score,
    compliancePercentage,
    onTimeCount,
    totalCount,
    missedCount,
    currentStreak,
    bestStreak,
    streakBrokenAt,
    protectedClaimsValue,
    atRiskValue,
    atRiskCount,
    activeCount,
    upcomingCount,
    display: formatDisplay(
      compliancePercentage,
      currentStreak,
      protectedClaimsValue
    ),
  };
}

// ── Get Score (cached) ────────────────────────────────────────────────────

export async function getComplianceScore(
  projectId: string
): Promise<ComplianceScoreResult> {
  // Get latest score, recalculate if stale (>1 hour)
  const latest = await prisma.complianceScore.findFirst({
    where: { projectId },
    orderBy: { calculatedAt: "desc" },
  });

  if (latest) {
    const oneHourAgo = new Date(Date.now() - 60 * 60 * 1000);
    if (latest.lastCalculatedAt > oneHourAgo) {
      return {
        projectId,
        score: latest.score,
        compliancePercentage: latest.totalCount > 0
          ? (latest.onTimeCount / latest.totalCount) * 100
          : null,
        onTimeCount: latest.onTimeCount,
        totalCount: latest.totalCount,
        missedCount: latest.missedCount,
        currentStreak: latest.currentStreak,
        bestStreak: latest.bestStreak,
        streakBrokenAt: latest.streakBrokenAt,
        protectedClaimsValue: Number(latest.protectedClaimsValue),
        atRiskValue: Number(latest.atRiskValue),
        atRiskCount: latest.atRiskCount,
        activeCount: latest.activeCount,
        upcomingCount: latest.upcomingCount,
        display: formatDisplay(
          latest.totalCount > 0
            ? (latest.onTimeCount / latest.totalCount) * 100
            : null,
          latest.currentStreak,
          Number(latest.protectedClaimsValue)
        ),
      };
    }
  }

  return calculateComplianceScore(projectId);
}

// ── Score History ─────────────────────────────────────────────────────────

export async function saveScoreSnapshot(
  projectId: string,
  periodType: "daily" | "weekly" | "monthly"
): Promise<void> {
  const score = await getComplianceScore(projectId);

  await prisma.complianceScoreHistory.upsert({
    where: {
      projectId_snapshotDate_periodType: {
        projectId,
        snapshotDate: startOfDay(new Date()),
        periodType,
      },
    },
    update: {
      compliancePercentage: score.compliancePercentage
        ? new Prisma.Decimal(score.compliancePercentage)
        : null,
      onTimeCount: score.onTimeCount,
      totalCount: score.totalCount,
      protectedClaimsValue: new Prisma.Decimal(score.protectedClaimsValue),
    },
    create: {
      projectId,
      snapshotDate: startOfDay(new Date()),
      compliancePercentage: score.compliancePercentage
        ? new Prisma.Decimal(score.compliancePercentage)
        : null,
      onTimeCount: score.onTimeCount,
      totalCount: score.totalCount,
      protectedClaimsValue: new Prisma.Decimal(score.protectedClaimsValue),
      periodType,
    },
  });
}

export async function getScoreHistory(
  projectId: string,
  period: "week" | "month" | "quarter" | "year"
) {
  const daysBack =
    period === "week"
      ? 7
      : period === "month"
        ? 30
        : period === "quarter"
          ? 90
          : 365;
  const startDate = new Date(Date.now() - daysBack * 24 * 60 * 60 * 1000);

  const history = await prisma.complianceScoreHistory.findMany({
    where: {
      projectId,
      snapshotDate: { gte: startDate },
    },
    orderBy: { snapshotDate: "asc" },
  });

  return history.map((h) => ({
    date: h.snapshotDate.toISOString().split("T")[0],
    compliancePercentage: h.compliancePercentage
      ? Number(h.compliancePercentage)
      : null,
    onTimeCount: h.onTimeCount,
    totalCount: h.totalCount,
    protectedClaimsValue: Number(h.protectedClaimsValue),
  }));
}

// ── Helpers ───────────────────────────────────────────────────────────────

function calculateStreak(
  notices: Array<{ onTimeStatus: boolean | null; sentAt: Date | null }>
): {
  currentStreak: number;
  bestStreak: number;
  streakBrokenAt: Date | null;
} {
  const sorted = [...notices]
    .filter((n) => n.sentAt)
    .sort((a, b) => b.sentAt!.getTime() - a.sentAt!.getTime());

  let currentStreak = 0;
  let streakBrokenAt: Date | null = null;

  for (const notice of sorted) {
    if (notice.onTimeStatus === true) {
      currentStreak++;
    } else {
      streakBrokenAt = notice.sentAt;
      break;
    }
  }

  // Best streak (chronological order)
  let bestStreak = 0;
  let tempStreak = 0;
  const chronological = [...sorted].reverse();
  for (const notice of chronological) {
    if (notice.onTimeStatus === true) {
      tempStreak++;
      bestStreak = Math.max(bestStreak, tempStreak);
    } else {
      tempStreak = 0;
    }
  }

  return { currentStreak, bestStreak, streakBrokenAt };
}

async function calculateProtectedClaimsValue(
  projectId: string
): Promise<number> {
  // Get on-time notices linked to deadlines with change event triggers
  const deadlines = await prisma.complianceDeadline.findMany({
    where: {
      projectId,
      status: "COMPLETED",
      triggerEventId: { not: null },
    },
    select: { triggerEventId: true },
  });

  if (deadlines.length === 0) {
    // Estimate based on on-time notice count
    const onTimeCount = await prisma.complianceNotice.count({
      where: { projectId, onTimeStatus: true },
    });
    return onTimeCount * 50000; // $50K average per notice (industry standard)
  }

  const eventIds = deadlines
    .map((d) => d.triggerEventId)
    .filter(Boolean) as string[];
  const events = await prisma.changeEvent.findMany({
    where: { id: { in: eventIds } },
    select: { estimatedValue: true },
  });

  const total = events.reduce(
    (sum, e) => sum + (e.estimatedValue ? Number(e.estimatedValue) : 50000),
    0
  );

  return total;
}

function getVerdict(percentage: number | null): string {
  if (percentage === null) return "No notices tracked yet";
  if (percentage >= 100) return "Excellent compliance";
  if (percentage >= 95) return "Strong compliance";
  if (percentage >= 90) return "Good compliance";
  if (percentage >= 80) return "Fair compliance — room for improvement";
  if (percentage >= 60) return "Poor compliance — action needed";
  return "Critical — immediate action required";
}

function formatCurrency(value: number): string {
  if (value >= 1_000_000) return `$${(value / 1_000_000).toFixed(1)}M`;
  if (value >= 1_000) return `$${(value / 1_000).toFixed(0)}K`;
  return `$${value.toFixed(0)}`;
}

function formatDisplay(
  compliancePercentage: number | null,
  currentStreak: number,
  protectedClaimsValue: number
) {
  return {
    scorePercentage:
      compliancePercentage !== null
        ? `${compliancePercentage.toFixed(0)}%`
        : "100%",
    streakDisplay:
      currentStreak > 0
        ? `${currentStreak} consecutive on time`
        : "No active streak",
    protectedValue: `${formatCurrency(protectedClaimsValue)} protected`,
    verdict: getVerdict(compliancePercentage),
  };
}

function startOfDay(date: Date): Date {
  const d = new Date(date);
  d.setHours(0, 0, 0, 0);
  return d;
}

File: backend/app/services/compliance/scoring.py
"""Compliance score calculation and history.

Calculates the compliance score based on notice delivery performance.
Formula: score = (onTimeCount / totalCount) * 100
Tracks streaks, claims values, and generates periodic snapshots.
"""

import logging
from datetime import datetime
from decimal import Decimal

from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.compliance import (
    ComplianceDeadline,
    ComplianceNotice,
    ComplianceScore,
    ComplianceScoreHistory,
)
from app.models.enums import (
    ComplianceNoticeStatus,
    DeadlineStatus,
    Severity,
)

logger = logging.getLogger(__name__)

# Default claims value per notice (used if no explicit value provided)
DEFAULT_CLAIMS_VALUE = Decimal("50000.00")


async def calculate_score(
    db: AsyncSession,
    project_id: str,
) -> ComplianceScore:
    """Calculate and upsert the compliance score for a project.

    Creates or updates the ComplianceScore record.
    """
    now = datetime.utcnow()

    # Count notices by on-time status
    sent_notices = await db.execute(
        select(ComplianceNotice).where(
            ComplianceNotice.project_id == project_id,
            ComplianceNotice.status.in_([
                ComplianceNoticeStatus.SENT,
                ComplianceNoticeStatus.ACKNOWLEDGED,
            ]),
        )
    )
    notices = sent_notices.scalars().all()

    total_count = len(notices)
    on_time_count = sum(1 for n in notices if n.on_time_status is True)
    missed_count = sum(1 for n in notices if n.on_time_status is False)

    # Score calculation
    score = round((on_time_count / total_count * 100)) if total_count > 0 else 100

    # Count active deadlines by severity
    active_deadlines = await db.execute(
        select(ComplianceDeadline).where(
            ComplianceDeadline.project_id == project_id,
            ComplianceDeadline.status.in_([
                DeadlineStatus.ACTIVE,
                DeadlineStatus.NOTICE_DRAFTED,
            ]),
        )
    )
    deadlines = active_deadlines.scalars().all()

    at_risk_count = sum(
        1 for d in deadlines
        if d.severity in (Severity.CRITICAL, Severity.WARNING)
    )
    active_count = len(deadlines)
    upcoming_count = sum(
        1 for d in deadlines
        if d.severity in (Severity.LOW, Severity.INFO)
    )

    # Claims value calculation
    protected_value = Decimal(on_time_count) * DEFAULT_CLAIMS_VALUE
    at_risk_value = Decimal(at_risk_count) * DEFAULT_CLAIMS_VALUE

    # Calculate streak
    streak = _calculate_streak(notices)

    # Upsert score record
    existing = await db.execute(
        select(ComplianceScore)
        .where(ComplianceScore.project_id == project_id)
        .order_by(ComplianceScore.calculated_at.desc())
        .limit(1)
    )
    score_record = existing.scalar_one_or_none()

    if score_record:
        old_streak = score_record.current_streak
        score_record.score = score
        score_record.on_time_count = on_time_count
        score_record.total_count = total_count
        score_record.missed_count = missed_count
        score_record.at_risk_count = at_risk_count
        score_record.active_count = active_count
        score_record.upcoming_count = upcoming_count
        score_record.protected_claims_value = protected_value
        score_record.at_risk_value = at_risk_value
        score_record.current_streak = streak
        score_record.best_streak = max(score_record.best_streak, streak)
        score_record.last_calculated_at = now
        score_record.calculated_at = now
        score_record.details = _build_details(
            score, on_time_count, total_count, missed_count,
            at_risk_count, active_count, streak,
        )

        # Track streak broken
        if streak < old_streak and old_streak > 0:
            score_record.streak_broken_at = now
    else:
        score_record = ComplianceScore(
            project_id=project_id,
            score=score,
            details=_build_details(
                score, on_time_count, total_count, missed_count,
                at_risk_count, active_count, streak,
            ),
            current_streak=streak,
            best_streak=streak,
            protected_claims_value=protected_value,
            at_risk_value=at_risk_value,
            on_time_count=on_time_count,
            total_count=total_count,
            missed_count=missed_count,
            at_risk_count=at_risk_count,
            active_count=active_count,
            upcoming_count=upcoming_count,
            last_calculated_at=now,
            calculated_at=now,
        )
        db.add(score_record)

    await db.flush()
    logger.info(
        "Score for project %s: %d%% (%d/%d on time, streak %d)",
        project_id, score, on_time_count, total_count, streak,
    )
    return score_record


async def create_score_snapshot(
    db: AsyncSession,
    project_id: str,
    period_type: str = "daily",
) -> ComplianceScoreHistory:
    """Create a point-in-time snapshot of the compliance score."""
    now = datetime.utcnow()

    # Get current score
    score = await calculate_score(db, project_id)

    # Count notices sent in the period (last 24h for daily, 7d for weekly)
    from datetime import timedelta
    period_hours = 24 if period_type == "daily" else 168
    period_start = now - timedelta(hours=period_hours)

    notices_in_period = await db.execute(
        select(func.count(ComplianceNotice.id)).where(
            ComplianceNotice.project_id == project_id,
            ComplianceNotice.sent_at >= period_start,
            ComplianceNotice.status.in_([
                ComplianceNoticeStatus.SENT,
                ComplianceNoticeStatus.ACKNOWLEDGED,
            ]),
        )
    )
    count = notices_in_period.scalar() or 0

    # Check for existing snapshot (upsert)
    existing = await db.execute(
        select(ComplianceScoreHistory).where(
            ComplianceScoreHistory.project_id == project_id,
            ComplianceScoreHistory.snapshot_date == now.replace(
                hour=0, minute=0, second=0, microsecond=0
            ),
            ComplianceScoreHistory.period_type == period_type,
        )
    )
    snapshot = existing.scalar_one_or_none()

    snapshot_date = now.replace(hour=0, minute=0, second=0, microsecond=0)
    percentage = Decimal(str(score.score))

    if snapshot:
        snapshot.compliance_percentage = percentage
        snapshot.on_time_count = score.on_time_count
        snapshot.total_count = score.total_count
        snapshot.notices_sent_in_period = count
        snapshot.protected_claims_value = score.protected_claims_value
    else:
        snapshot = ComplianceScoreHistory(
            project_id=project_id,
            snapshot_date=snapshot_date,
            compliance_percentage=percentage,
            on_time_count=score.on_time_count,
            total_count=score.total_count,
            notices_sent_in_period=count,
            protected_claims_value=score.protected_claims_value,
            period_type=period_type,
        )
        db.add(snapshot)

    await db.flush()
    return snapshot


async def get_score_history(
    db: AsyncSession,
    project_id: str,
    period_type: str = "daily",
    limit: int = 30,
) -> list[ComplianceScoreHistory]:
    """Get compliance score history for trending."""
    result = await db.execute(
        select(ComplianceScoreHistory)
        .where(
            ComplianceScoreHistory.project_id == project_id,
            ComplianceScoreHistory.period_type == period_type,
        )
        .order_by(ComplianceScoreHistory.snapshot_date.desc())
        .limit(limit)
    )
    return list(result.scalars().all())


def _calculate_streak(notices: list[ComplianceNotice]) -> int:
    """Calculate current consecutive on-time streak.

    Sorts notices by sent_at descending and counts consecutive on-time notices.
    """
    if not notices:
        return 0

    sorted_notices = sorted(
        [n for n in notices if n.sent_at],
        key=lambda n: n.sent_at,
        reverse=True,
    )

    streak = 0
    for notice in sorted_notices:
        if notice.on_time_status is True:
            streak += 1
        else:
            break

    return streak


def _build_details(
    score: int,
    on_time: int,
    total: int,
    missed: int,
    at_risk: int,
    active: int,
    streak: int,
) -> dict:
    """Build the JSON details blob for the score record."""
    return {
        "score": score,
        "onTimeCount": on_time,
        "totalCount": total,
        "missedCount": missed,
        "atRiskCount": at_risk,
        "activeDeadlines": active,
        "currentStreak": streak,
        "formula": "onTimeCount / totalCount * 100",
    }
